test/icinga-notification.cpp

```diff
@@ -1,11 +1,58 @@
 /* Icinga 2 | (c) 2012 Icinga GmbH | GPLv2+ */
 
+#include "icinga/host.hpp"
 #include "icinga/notification.hpp"
+#include "icinga/notificationcommand.hpp"
+#include "icinga/service.hpp"
+#include "icinga/user.hpp"
 #include <BoostTestTargetConfig.h>
 #include <iostream>
 
 using namespace icinga;
 
+struct DuplicateDueToFilterHelper
+{
+	Host::Ptr h = new Host();
+	Service::Ptr s = new Service();
+	User::Ptr u = new User();
+	NotificationCommand::Ptr nc = new NotificationCommand();
+	Notification::Ptr n = new Notification();
+	unsigned int called = 0;
+
+	DuplicateDueToFilterHelper()
+	{
+		h->SetName("example.com", true);
+		h->Register();
+
+		s->SetShortName("disk", true);
+		h->AddService(s);
+
+		u->SetName("jdoe", true);
+		u->SetTypeFilter(~0);
+		u->SetStateFilter(~0);
+		u->Register();
+
+		nc->SetName("mail", true);
+		nc->SetExecute(new Function("", [this]() { ++called; }), true);
+		nc->Register();
+
+		n->SetFieldByName("host_name", "example.com", false, DebugInfo());
+		n->SetFieldByName("service_name", "disk", false, DebugInfo());
+		n->SetFieldByName("command", "mail", false, DebugInfo());
+		n->SetUsersRaw(new Array({"jdoe"}), true);
+		n->SetTypeFilter(~NotificationRecovery);
```

Comment:
To meet @lippserd's explicit requirements:

* Crit(sent) -> Warn(filtered) -> Crit(not sent)
* Crit(sent) -> OK(filtered) -> Crit(sent)

Reply:
I only remember that we discussed that it would be easiest to reset state once recovery notifications have been sent. Anyway, for me it makes sense to reset state, even if the recovery notifications are filtered away.